Load rustdoc's JS search index on-demand.

jsha opened the pull request:
Instead of being loaded on every page, the JS search index is now loaded when either (a) there is a `?search=` param, or (b) the search input is focused.

This saves both CPU and bandwidth. As of Feb 2021, https://doc.rust-lang.org/search-index1.50.0.js is 273,838 bytes gzipped or 2,544,939 bytes uncompressed. Evaluating it takes 445 ms of CPU time in Chrome 88 on a i7-10710U CPU (out of a total ~2,100 ms page reload).

Tested on Firefox and Chrome.

New:
https://jacob.hoffman-andrews.com/rust/search-on-demand/std/primitive.slice.html
https://jacob.hoffman-andrews.com/rust/search-on-demand/std/primitive.slice.html?search=fn

Old:
https://jacob.hoffman-andrews.com/rust/search-on-load/std/primitive.slice.html
https://jacob.hoffman-andrews.com/rust/search-on-load/std/primitive.slice.html?search=fn

rust-highfive commented:
Some changes occurred in HTML/CSS/JS.

cc @GuillaumeGomez

rust-highfive commented:
Thanks for the pull request, and welcome! The Rust team is excited to review your changes, and you should hear from @GuillaumeGomez (or someone else) soon.

If any changes to this PR are deemed necessary, please add them as extra commits. This ensures that the reviewer can see what has changed since they last reviewed the code. Due to the way GitHub handles out-of-date commits, this should also make it reasonably obvious what issues have or haven't been addressed. Large or tricky changes may require several passes of review and changes.

Please see [the contribution instructions](https://rustc-dev-guide.rust-lang.org/contributing.html) for more information.

GuillaumeGomez commented:
I have two problems with this change:
 1. The crates list for the search isn't loaded anymore until you focus on the search input.
 2. Your browser is heavily using caching, so once it loaded the search index once, it's never loaded anymore until the hash/URL of the URL changes.

jsha commented:
> The crates list for the search isn't loaded anymore until you focus on the search input.

Pushed a fix. Thanks for spotting. I've also updated the demo pages at https://jacob.hoffman-andrews.com/rust/search-on-demand/std/primitive.slice.html.

> Your browser is heavily using caching, so once it loaded the search index once, it's never loaded anymore until the hash/URL of the URL changes.

In my tests, the browser is spending significant time parsing the JS (and the JSON inside). That's time that blocks other JS work, like building the toggles, which in turn blocks the overall page load from completing. This technique also reduces the memory footprint of doc pages in the common case where someone hasn't used search.

GuillaumeGomez commented:
When I pick a crate to filter the search results, the picked crate won't show up when I reload the page until I load the search index.

GuillaumeGomez commented:
Another thing which bothers me is the fact that the crate list arrives quite a while **after** clicking on the menu. It's really bad from a UX perspective...

jsha commented:
> When I pick a crate to filter the search results, the picked crate won't show up when I reload the page until I load the search index.

Good catch, fixed (and updated the demo pages).

> Another thing which bothers me is the fact that the crate list arrives quite a while after clicking on the menu. It's really bad from a UX perspective...

Agreed, this also bothers me. Perhaps it's possible to generate the crate list statically rather than in JS? I'll take a look.

By the way, this noticeable delay is a good demonstration that loading the search JSON takes non-negligible time even when the JS is cached.

GuillaumeGomez commented:
> By the way, this noticeable delay is a good demonstration that loading the search JSON takes non-negligible time even when the JS is cached.

That's actually a great argument in favor for this change!

> > Another thing which bothers me is the fact that the crate list arrives quite a while after clicking on the menu. It's really bad from a UX perspective...
> Agreed, this also bothers me. Perhaps it's possible to generate the crate list statically rather than in JS? I'll take a look.

It'd require another JS file to store the crate list but it's possible.

In the last version on your website, it seems like the search-index is now loaded every time (even with an empty local storage).

jsha commented:
> It'd require another JS file to store the crate list but it's possible.

Is the crate list not known at generation time for any particular page? Also, are there situations where the crate list is quite large, or is the situation on `doc.rust-lang.org` representative (5 crates)?

> In the last version on your website, it seems like the search-index is now loaded every time (even with an empty local storage).

Yep. The issue here is that if you pick a crate, then go back to "All crates", there's still a storage entry. Updated the code to ignore "All crates" (and pushed demo).

GuillaumeGomez commented:
> Is the crate list not known at generation time for any particular page? Also, are there situations where the crate list is quite large, or is the situation on `doc.rust-lang.org` representative (5 crates)?

It's actually **very** common to have more than 5 crates. To be exact, it depends on how many dependencies you have (so I'll let you do the maths :p).

> Yep. The issue here is that if you pick a crate, then go back to "All crates", there's still a storage entry. Updated the code to ignore "All crates" (and pushed demo).

I precised that my local storage was empty, which meant that I removed the local storage "rustdoc-saved-filter-crate" entry completely through the web dev tools. ;)

GuillaumeGomez commented:
Sorry, forgot to answer to one of your questions:

> Is the crate list not known at generation time for any particular page? 

No, and the reason is actually a bit more complex than what it seems: you might add more crates to your local documentation but the search index has to work in any case and not simply be overwritten. So to do so, if the search-index already has this crate, it overwrites this entry specifically, otherwise it happens it.

jsha commented:
I pushed a new version (and updated demo) that generates a separate, smaller, crates.js listing the crates. This is used for building the dropdown and also for adding sidebar items depending on the page. There's still some issue: my crates.js has `alloc`, `core`, `proc_macro`, and `std`. But if I look on doc.rust-lang.org, the list should also include `test`.

> > In the last version on your website, it seems like the search-index is now loaded every time (even with an empty local storage).
> I precised that my local storage was empty, which meant that I removed the local storage "rustdoc-saved-filter-crate" entry completely through the web dev tools. ;)

Found the issue. Looks like a difference between Firefox and Chrome. `savedCrate` was `undefined` in Chrome, but `null` in Firefox.

GuillaumeGomez commented on src/librustdoc/html/render/mod.rs:
```diff
@@ -1048,11 +1048,15 @@ themePicker.onblur = handleThemeButtonsBlur;
+
+    let crate_list_dst = cx.dst.join(&format!("crates{}.js", cx.shared.resource_suffix));
+    let crate_list =
+        format!("const crates = [{}];", krates.iter().map(|k| format!("\"{}\"", k)).join(","));
```
`const` isn't part of ES5, please use instead: `window.CRATES = [{}]`. Also: this update is problematic because you remove crates that aren't listed, which is why you don't have `test` listed in the crates. Please update the list instead of rewriting it completely.

GuillaumeGomez replied:
In case you wonder: it's why we update the search index like that and have a crate define on each line separately.

GuillaumeGomez commented on src/librustdoc/html/static/main.js:
```diff
@@ -2952,6 +2937,35 @@ function defocusSearchBar() {
+    }
+
+    // Assumes `crates{version}.js` has already been loaded, setting the global const `crates`.
+    addSearchOptions(crates);
```
I don't like much to "assume" things are there in an async world. Instead, please call the function to init the crates lists from the `crates{version}.js` file directly.

GuillaumeGomez replied:
Ok nevermind. It's fine as is. However, please use `window.crates` or equivalent. I don't like the usage of global with such common names, it makes the readibility a lot more problematic.

GuillaumeGomez commented on src/librustdoc/html/static/main.js:
```diff
@@ -2952,6 +2937,35 @@ function defocusSearchBar() {
+    // Assumes `crates{version}.js` has already been loaded, setting the global const `crates`.
+    addSearchOptions(crates);
+    addSidebarCrates(crates);
+    var crateSearchDropDown = document.getElementById("crate-search");
```
Can you put all this in a function too so we can avoid creating variables in the "global" scope of this function please?

GuillaumeGomez commented:
I think we're getting close to something good! Not having to load the search-index when unneeded is a good idea. I'm still worried about the big delay we have on the first search though... I'll think about that.

jsha commented:
A couple of thoughts on reducing the feeling of delay on the first search:

Right now, there's no reaction on the rest of the page when you highlight the search box. It would be nice to insert some text below the search box like "Getting ready to search." Alternately: right now the search box says "Click or press 'S' to search, '?' for more options." And it continues to say that even after you focus the search box. We could make it so focusing the search box changes this to say "Type in your search here." That would indicate to the user that their input was received.

Similarly, when the user starts typing, we should be responsive to that typing even if the index is not yet loaded. For instance, We could add text under the search box "Searching for 'yourquery'." One problem: the index JS load is one big event, not interruptible by other events. We could probably work around that by breaking up the JSON.parse into multiple calls (one for each crate), and adding setTimeout between. Similarly, I think there's a fair amount of work happening to turn the parsed JSON into in-memory structures; we could similarly add setTimeout during that work to make it possible to update the page in between.

rust-log-analyzer commented:
The job **`mingw-check`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/runs/1948357967) [(plain)](https://github.com/rust-lang/rust/commit/80f1231cf49cf0ac10df27df37a2e18ac0363f8b/checks/1948357967/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
configure: rust.channel         := nightly
configure: rust.debug-assertions := True
configure: llvm.assertions      := True
configure: dist.missing-tools   := True
configure: build.configure-args := ['--enable-sccache', '--disable-manage-submodu ...
configure: writing `config.toml` in current directory
configure: 
configure: run `python /checkout/x.py --help`
configure: 
---
Build completed successfully in 0:00:55
+ /scripts/validate-toolstate.sh
Cloning into 'rust-toolstate'...
<Nothing changed>
+ es-check es5 ../src/librustdoc/html/static/main.js ../src/librustdoc/html/static/settings.js ../src/librustdoc/html/static/source-script.js ../src/librustdoc/html/static/storage.js
ES-Check: there were 1 ES version matching errors.
          ES-Check Error:
          ----
          · erroring file: ../src/librustdoc/html/static/main.js
          · erroring file: ../src/librustdoc/html/static/main.js
          · error: SyntaxError: The keyword 'const' is reserved (2939:8)
          · see the printed err.stack below for context


          SyntaxError: The keyword 'const' is reserved (2939:8)
    at Parser.pp$4.raise (/node-v14.4.0-linux-x64/lib/node_modules/es-check/node_modules/acorn/dist/acorn.js:2825:15)
    at Parser.pp$3.checkUnreserved (/node-v14.4.0-linux-x64/lib/node_modules/es-check/node_modules/acorn/dist/acorn.js:2752:12)
    at Parser.pp$3.parseIdent (/node-v14.4.0-linux-x64/lib/node_modules/es-check/node_modules/acorn/dist/acorn.js:2781:12)
    at Parser.pp$3.parseExprAtom (/node-v14.4.0-linux-x64/lib/node_modules/es-check/node_modules/acorn/dist/acorn.js:2198:21)
    at Parser.pp$3.parseExprSubscripts (/node-v14.4.0-linux-x64/lib/node_modules/es-check/node_modules/acorn/dist/acorn.js:2089:21)
    at Parser.pp$3.parseMaybeUnary (/node-v14.4.0-linux-x64/lib/node_modules/es-check/node_modules/acorn/dist/acorn.js:2066:19)
    at Parser.pp$3.parseExprOps (/node-v14.4.0-linux-x64/lib/node_modules/es-check/node_modules/acorn/dist/acorn.js:2010:21)
    at Parser.pp$3.parseMaybeConditional (/node-v14.4.0-linux-x64/lib/node_modules/es-check/node_modules/acorn/dist/acorn.js:1993:21)
    at Parser.pp$3.parseMaybeAssign (/node-v14.4.0-linux-x64/lib/node_modules/es-check/node_modules/acorn/dist/acorn.js:1968:21)
    at Parser.pp$3.parseExpression (/node-v14.4.0-linux-x64/lib/node_modules/es-check/node_modules/acorn/dist/acorn.js:1933:21)
    at Parser.pp$1.parseStatement (/node-v14.4.0-linux-x64/lib/node_modules/es-check/node_modules/acorn/dist/acorn.js:877:47)
    at Parser.pp$1.parseBlock (/node-v14.4.0-linux-x64/lib/node_modules/es-check/node_modules/acorn/dist/acorn.js:1161:23)
    at Parser.pp$3.parseFunctionBody (/node-v14.4.0-linux-x64/lib/node_modules/es-check/node_modules/acorn/dist/acorn.js:2671:24)
    at Parser.pp$1.parseFunction (/node-v14.4.0-linux-x64/lib/node_modules/es-check/node_modules/acorn/dist/acorn.js:1283:10)
    at Parser.pp$1.parseFunctionStatement (/node-v14.4.0-linux-x64/lib/node_modules/es-check/node_modules/acorn/dist/acorn.js:983:17)
    at Parser.pp$1.parseStatement (/node-v14.4.0-linux-x64/lib/node_modules/es-check/node_modules/acorn/dist/acorn.js:830:19)
```

</details>

jsha commented:
By the way, I was looking at the delay when searching and realized we have an artificial delay of 500ms before any search gets executed: https://github.com/rust-lang/rust/blob/301ad8a4fa3ea56fb980443b7997c8f9d72dd717/src/librustdoc/html/static/main.js#L1935.

GuillaumeGomez commented:
Yep, but in here we outrun this timeout apparently. ;)

jsha commented:
I tested the `setTimeout` concept. Unfortunately it didn't seem to help. None of the keyboard input events fired between loading, parsing, and buildIndex.

I did find that `search` was running twice. Once, right after `buildIndex`, it would run with whatever was in the search box, even if I was still typing. Since the search itself is a bit expensive (250ms on my machine), doing it twice slows everything down. I fixed that so that "right after buildIndex" search only fires when there's a search in the URL bar.

Overall, I think this is the best we're likely to do. If we wanted to get really fancy, we could use a Web Worker that always has a copy of the index loaded. That would make searches blazing fast, but it wouldn't work for local files so we'd need two code paths. That's probably too much complexity.

GuillaumeGomez commented on src/librustdoc/html/render/mod.rs:
```diff
@@ -1037,22 +1037,27 @@ themePicker.onblur = handleThemeButtonsBlur;
+
+    let crate_list_dst = cx.dst.join(&format!("crates{}.js", cx.shared.resource_suffix));
+    let crate_list =
+        format!("window.ALL_CRATES = [{}];", krates.iter().map(|k| format!("\"{}\"", k)).join(","));
```
I think this is still invalid: you'll overwrite any crate that isn't a dependency of the current one (when you are in a workspace for example, or simply documenting unrelated code in the same location). You need to extract the crate list and then update it like we do in the search index (we don't overwrite, we update, which is very different!).

jsha replied:
I may be misunderstanding, but I think this does the right thing. the `krates` list comes from `collect_json`, which reads it from the `search-index{}.js` as it currently stands. So as each run of `rustdoc` happens, new crates are added to `search-index{}.js`, and `crates{}.js` is updated based on the latest contents of `search-index{}.js`.

jsha replied:
As a test, I checked out https://github.com/servo/rust-url, which is a workspace containing 5 members:

```
[workspace]
members = ["url", "form_urlencoded", "idna", "percent_encoding", "data-url"]
```

Building docs with this branch produces this (truncated) search-index:

```
var searchIndex = JSON.parse('{\
"data_url":{"doc":"Processing of <code>data:</code> URLs according to the Fetch 
"form_urlencoded":{"doc":"Parser and serializer for the …","i":[[5,"parse","fo
"idna":{"doc":"This Rust crate implements IDNA per the WHATWG URL …","i":[[0,"
"matches":{"doc":"","i":[[14,"matches","matches","Check if an expression matches
"percent_encoding":{"doc":"URLs use special chacters to indicate the parts of th
"tinyvec":{"doc":"<code>tinyvec</code> provides 100% safe vec-like data structur
"tinyvec_macros":{"doc":"","i":[[14,"impl_mirrored","tinyvec_macros","",null,nul
"unicode_bidi":{"doc":"This crate implements the Unicode Bidirectional Algorithm
"unicode_normalization":{"doc":"Unicode character composition and decomposition 
"url":{"doc":"rust-url is an implementation of the URL Standard for the …","i"
}');
```

And this crates.js:
```
window.ALL_CRATES = ["data_url","form_urlencoded","idna","matches","percent_encoding","tinyvec","tinyvec_macros","unicode_bidi","unicode_normalization","url"];
```

Building docs with nightly produces this (truncated) search-index:

```var searchIndex = JSON.parse('{\
"data_url":{"doc":"Processing of <code>data:</code> URLs according to the Fetch 
"form_urlencoded":{"doc":"Parser and serializer for the …","i":[[5,"parse","fo
"idna":{"doc":"This Rust crate implements IDNA per the WHATWG URL …","i":[[0,"
"matches":{"doc":"","i":[[14,"matches","matches","Check if an expression matches
"percent_encoding":{"doc":"URLs use special chacters to indicate the parts of th
"tinyvec":{"doc":"<code>tinyvec</code> provides 100% safe vec-like data structur
"tinyvec_macros":{"doc":"","i":[[14,"impl_mirrored","tinyvec_macros","",null,nul
"unicode_bidi":{"doc":"This crate implements the Unicode Bidirectional Algorithm
"unicode_normalization":{"doc":"Unicode character composition and decomposition 
"url":{"doc":"rust-url is an implementation of the URL Standard for the …","i"
}');
```

GuillaumeGomez replied:
And what happens if you now document another non-related crate and make the output into this folder? The `search-index.js` will be updated and the `crates.js` will very likely only contain this new crate.

jsha replied:
This also works. I just tested by doing:

```
cd rust/ureq
cargo doc
cd ../conduit-cookie
cargo doc --target-dir ../ureq/target/
```

The resulting crates.js includes ureq, conduit-cookie, and all the dependencies of each:

```
window.ALL_CRATES = ["aead","aes","aes_gcm","aes_soft","base64","block_buffer","bytes","cfg_if","chunked_transfer","cipher","conduit","conduit_cookie","conduit_middleware","const_fn","cookie","cpuid_bool","crypto_mac","ctr","digest","fnv","form_urlencoded","generic_array","getrandom","ghash","hkdf","hmac","http","idna","itoa","libc","log","matches","once_cell","opaque_debug","percent_encoding","polyval","ppv_lite86","proc_macro2","proc_macro_hack","quote","rand","rand_chacha","rand_core","ring","rustls","sct","sha2","spin","standback","subtle","syn","time","time_macros","time_macros_impl","tinyvec","tinyvec_macros","typenum","unicode_bidi","unicode_normalization","unicode_xid","universal_hash","untrusted","ureq","url","webpki","webpki_roots"];
```

GuillaumeGomez replied:
I just checked, you are absolutely right!

GuillaumeGomez commented on src/librustdoc/html/static/main.js:
```diff
@@ -1917,8 +1918,8 @@ function defocusSearchBar() {
-        function startSearch() {
-            var callback = function() {
+        function registerSearchEvents() {
+            var searchAfter500ms = function() {
```
That's a nice naming improvement, thanks!

GuillaumeGomez commented on src/librustdoc/html/static/main.js:
```diff
@@ -2843,22 +2839,14 @@ function defocusSearchBar() {
-            }
-        }
-        crates_text.sort(function(a, b) {
+        crates.sort(function(a, b) {
```
I just thought: now that the crate list is separated from the search index, it might be nice to sort them when generating the crate list in the rust code instead of doing it client side every time. What do you think? If you think it's too much, please add a FIXME comment so it can be taken care of later on.

jsha replied:
Yep, I'd be happy to do this!

GuillaumeGomez commented on src/librustdoc/html/static/main.js:
```diff
@@ -2950,8 +2938,46 @@ function defocusSearchBar() {
+            }
+        }
+
+        // Assumes `crates{version}.js` has already been loaded, setting window.ALL_CRATES.
```
```suggestion
        // `crates{version}.js` should always be loaded before this script, so we can use it safely. Time to set `window.ALL_CRATES`!
```

GuillaumeGomez commented on src/librustdoc/html/static/main.js:
```diff
@@ -2950,8 +2938,46 @@ function defocusSearchBar() {
+
+        search_input.addEventListener("focus", function(){
+            search_input.origPlaceholder = search_input.placeholder;
+            search_input.placeholder = "Type your search here.";
```
I'm not sure to understand why you're updating the placeholder when the input gets the focus. Any reason?

jsha replied:
This is part of my attempts to make the search feel more responsive. Even though it might take some time for the search index to load, and for your query to run, it's nice for the page to immediately respond to your input in some way. This is the "response" idea from [Response, Animation, Idle, Load (RAIL)](https://web.dev/rail/).

GuillaumeGomez replied:
Tested it and I actually like how it reacts so let's keep it! :)

GuillaumeGomez commented on src/librustdoc/html/static/main.js:
```diff
@@ -2950,8 +2938,46 @@ function defocusSearchBar() {
+        addSearchOptions(window.ALL_CRATES);
+        addSidebarCrates(window.ALL_CRATES);
+
+        search_input.addEventListener("focus", function(){
```
```suggestion
        search_input.addEventListener("focus", function() {
```

GuillaumeGomez commented on src/librustdoc/html/static/main.js:
```diff
@@ -2950,8 +2938,46 @@ function defocusSearchBar() {
+            search_input.placeholder = "Type your search here.";
+            loadSearch();
+        });
+        search_input.addEventListener("blur", function(){
```
```suggestion
        search_input.addEventListener("blur", function() {
```

GuillaumeGomez replied:
Still need to add the missing whitespace. ;)

GuillaumeGomez commented on src/librustdoc/html/static/main.js:
```diff
@@ -2950,8 +2938,46 @@ function defocusSearchBar() {
+            loadSearch();
+        });
+        search_input.addEventListener("blur", function(){
+            search_input.placeholder = search_input.origPlaceholder;
```
If replacing the placeholder isn't useful, this callback can be removed.

rust-log-analyzer commented:
The job **`mingw-check`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/runs/1990442089) [(plain)](https://github.com/rust-lang/rust/commit/9d52e53056c4aec186ced518e35f1f1f19ef429b/checks/1990442089/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
configure: rust.channel         := nightly
configure: rust.debug-assertions := True
configure: llvm.assertions      := True
configure: dist.missing-tools   := True
configure: build.configure-args := ['--enable-sccache', '--disable-manage-submodu ...
configure: writing `config.toml` in current directory
configure: 
configure: run `python /checkout/x.py --help`
configure: 
---
Checking which error codes lack tests...
Found 436 error codes
Found 0 error codes with no tests
Done!
tidy error: /checkout/src/librustdoc/html/static/main.js:2961: line longer than 100 chars
some tidy checks failed

command did not execute successfully: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-tools-bin/tidy" "/checkout" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/cargo" "/checkout/obj/build"
expected success, got: exit code: 1

```

</details>

bors commented:
:umbrella: The latest upstream changes (presumably #82511) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#conflicts).

GuillaumeGomez commented on src/librustdoc/html/render/mod.rs:
```diff
@@ -1039,22 +1039,29 @@ themePicker.onblur = handleThemeButtonsBlur;
     all_indexes.push(search_index);
+    krates.push(krate.name.to_string());
+    krates.sort();
+    krates.dedup();
```
Just wondering: is that case even possible? Normally you can only have each crate once. Did you found a case where there was a duplicate?

jsha replied:
I was moving this up from here: https://github.com/rust-lang/rust/pull/82310/files#diff-40a0eb025da61717b3b765ceb7fab21d91af3012360e90b9f46e15a4047946faL1081. I do think it's possible that `krates.push(krate.name.to_string())` adds a crate name that is a duplicate of a crate already in the list.

GuillaumeGomez replied:
Look at the code of `collect_json`, this more precisely:

```rust
// ...
let prefix = format!("\"{}\"", krate);
// ...
    if line.starts_with(&prefix) {
        continue;
    }
// ...
```

If the crate is already in the search index, we don't add it to the crates list returned. So the `dedup` is unnecessary here.

GuillaumeGomez commented:
I went through the changes once again and I think we're getting close to the approval. Just two comments (one nit and one question about the `dedup`). Thanks for going through this review process! :)

jsha commented:
Updated, thanks!

GuillaumeGomez commented:
Thanks!

@bors: r+

bors commented:
:pushpin: Commit 768d5e950953738a54480e530341964838d29da2 has been approved by `GuillaumeGomez`

<!-- @bors r=GuillaumeGomez 768d5e950953738a54480e530341964838d29da2 -->
<!-- homu: {"type":"Approved","sha":"768d5e950953738a54480e530341964838d29da2","approver":"GuillaumeGomez"} -->
